compat/mingw.c

```diff
@@ -426,6 +428,47 @@ static wchar_t *normalize_ntpath(wchar_t *wbuf)
 	return wbuf;
 }
 
+/* rename and unlink a file. This is needed for the ability to recreate a file
+ * with the same name, for example on checkout which unlinks and recreates the file */
+static int rename_and_unlink(const wchar_t *pathname, int (*unlinker)(const wchar_t *))
+{
+	wchar_t temppath[MAX_LONG_PATH];
+	int tries = 0;
+	*temppath = 0;
+	/* leave some space for the .unlink_XXXXXX extension */
+	wcsncat(temppath, pathname, sizeof(temppath) - 15);
+	wcsncat(temppath, L".unlink_XXXXXX", sizeof(temppath));
+	if (!_wmktemp(temppath))
+		return unlinker(pathname);
+	do {
+		if (MoveFileExW(pathname, temppath, 0)) {
+			const int res = unlinker(temppath);
+			if (res) {
+				const DWORD err = GetLastError();
+				MoveFileExW(temppath, pathname, 0);
+				SetLastError(err);
+			}
+			return res;
+		}
+	} while (retry_ask_yes_no(&tries, ""));
```

Comment:
Here is some pseudo code for what I have in mind. It implements `mingw_unlink` below (new line 472).

```
mingw_unlink(pathname) {
  move_to_temporary(pathname, &tempPath);
  if (failed)
    return with error status;
  do_unlink(tempPath);
}
```

The `do_unlink` function is equivalent to the original `mingw_unlink` function. It contains the retry loop for unlinking but does not concern itself with renaming. The `move_to_temporary` function, on the other hand does not concern itself with unlinking. It will try to rename the file, maybe asking the user to retry if it fails. It the user says no, we return with an error and `mingw_unlink` aborts.

In particular, if the rename succeeds and unlink does not, then we leave the renamed file in the work tree. There is precedent for leaving untracked files in the working tree, for example during merge conflicts or when using third party merge tools.

Reply:
Furthermore, please do not try to unlink the original file when the rename fails. This step will succeed, but we will fail later when we create a new file with the same name. This is exactly the behavior we want to fix.

Reply:
> My suggestion is to do the rename in mingw_unlink

But we only want to do the rename if the file is in use. Otherwise it becomes too slow. I removed the internal retry loop (I think I added it to test if it helps on a particular case, which it didn't).

> please do not try to unlink the original file when the rename fails.

You're right. Done.